compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs

```diff
@@ -313,12 +315,122 @@ getTypeClassDictionary world = \case
     _ ->
         Nothing
 
-simplifyExpr :: World -> Expr -> Expr
-simplifyExpr world = fst . cata go
+-- | Go inside a closed expression and try to inline and beta reduce
+-- types and arguments, to try monomorphizing.
+inlineClosedExpr :: Int -> World -> Expr -> Simplifier Expr
+inlineClosedExpr d world e = do
+    e' <- go e
+    if e == e'
+        then pure e
+        else simplifyExpr' (d-1) e'
+  where
+    go = \case
+        e@(EVal x) -> do
+            case lookupValue x world of
+                Left _ -> pure e
+                Right dval -> pure (dvalBody dval)
+
+        ETmApp e1 e2 -> do
+            e1' <- go e1
+            case stripLoc e1' of
+                ETmLam (x,_) b -> do
+                    pure (applySubstInExpr (exprSubst x e2) b)
```

Comment:
Can this not blow up if the variable is used very often in the body of the lambda?

Reply:
You're right, this should probably be a let.